### PR TITLE
Update packet function 

### PR DESCRIPTION
**Pull request by dgsbl**

### CODE REVIEW

File: modules/tibc/core/04-packet/keeper/keeper.go [hunk 1]
```diff
@@ -305,3 +331,64 @@ func (k Keeper) iterateHashes(_ sdk.Context, iterator db.Iterator, cb func(sourc
...
+	for ; iterator.Valid(); iterator.Next() {
+		seq := binary.BigEndian.Uint64(iterator.Key())
+		if seq <= sequence && !k.HasPacketCommitment(ctx, sourceChain, destChain, seq){
+			return sdkerrors.Wrapf(types.ErrInvalidCleanPacket, "packet with sequence %d has not been ack", seq)
+		}
+	}
+	return nil
+}
+
+func (k Keeper) cleanPacketCommitmentBySeq(ctx sdk.Context, sourceChain, destChain string, sequence uint64) {
```
Comment (chengwenxi) [inline modules/tibc/core/04-packet/keeper/keeper.go:348]: Clean packet commitment on receiving ack packet.

File: modules/tibc/core/04-packet/keeper/keeper.go [hunk 2]
```diff
@@ -305,3 +331,64 @@ func (k Keeper) iterateHashes(_ sdk.Context, iterator db.Iterator, cb func(sourc
...
 	}
 }
+
+func (k Keeper) ValidateCleanPacket(ctx sdk.Context, sourceChain, destChain string, sequence uint64) error{
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte(host.PacketCommitmentPrefixPath(sourceChain, destChain)))
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		seq := binary.BigEndian.Uint64(iterator.Key())
+		if seq <= sequence && !k.HasPacketCommitment(ctx, sourceChain, destChain, seq){
```
Comment (chengwenxi) [inline modules/tibc/core/04-packet/keeper/keeper.go:341]: check ack commitment

Comment (chengwenxi) [reply modules/tibc/core/04-packet/keeper/keeper.go:341]:
```go
if seq > sequence {
    break
}
```

File: modules/tibc/core/04-packet/keeper/keeper.go [hunk 3]
```diff
@@ -305,3 +331,64 @@ func (k Keeper) iterateHashes(_ sdk.Context, iterator db.Iterator, cb func(sourc
...
+	}
+}
+
+func (k Keeper) cleanPacketAcknowledgementBySeq(ctx sdk.Context, sourceChain, destChain string, sequence uint64) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte(host.PacketAcknowledgementPrefixPath(sourceChain, destChain)))
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		seq := binary.BigEndian.Uint64(iterator.Key())
+		if seq <= sequence {
```
Comment (chengwenxi) [inline modules/tibc/core/04-packet/keeper/keeper.go:366]: ditto.

File: modules/tibc/core/04-packet/keeper/keeper.go [hunk 4]
```diff
@@ -305,3 +331,64 @@ func (k Keeper) iterateHashes(_ sdk.Context, iterator db.Iterator, cb func(sourc
...
+	}
+}
+
+func (k Keeper) cleanReceiptBySeq(ctx sdk.Context, sourceChain, destChain string, sequence uint64) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte(host.PacketReceiptPrefixPath(sourceChain, destChain)))
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		seq := binary.BigEndian.Uint64(iterator.Key())
+		if seq <= sequence {
```
Comment (chengwenxi) [inline modules/tibc/core/04-packet/keeper/keeper.go:378]: ditto.

File: modules/tibc/core/04-packet/keeper/keeper.go [hunk 5]
```diff
@@ -305,3 +331,64 @@ func (k Keeper) iterateHashes(_ sdk.Context, iterator db.Iterator, cb func(sourc
...
+			k.deletePacketReceipt(ctx, sourceChain, destChain, seq)
+		}
+	}
+}
+
+func (k Keeper) cleanCleanPacketCommitmentBySeq(ctx sdk.Context, sourceChain, destChain string, sequence uint64) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte(host.CleanPacketCommitmentPrefixPath(sourceChain, destChain)))
```
Comment (chengwenxi) [inline modules/tibc/core/04-packet/keeper/keeper.go:386]: Store `CleanPacket` as
```sheel
cleans/{sourceChain}/{destChain} -> sequence
```

File: modules/tibc/core/04-packet/keeper/keeper.go [hunk 6]
```diff
@@ -305,3 +325,19 @@ func (k Keeper) iterateHashes(_ sdk.Context, iterator db.Iterator, cb func(sourc
...
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		seq := binary.BigEndian.Uint64(iterator.Key())
+		if sequence < seq{
+			break
+		}
+		k.deletePacketAcknowledgement(ctx, sourceChain, destChain, seq)
+		if k.HasPacketReceipt(ctx, sourceChain, destChain, seq){
+			k.deletePacketReceipt(ctx, sourceChain, destChain, seq)
+		}
```
Comment (chengwenxi) [inline modules/tibc/core/04-packet/keeper/keeper.go:341]: how about handling it as
```go
ackIterator := store.ReverseIterator(store, host.PacketAcknowledgementKey(sourceChain, destChain, 0), host.PacketAcknowledgementKey(sourceChain, destChain, sequence+1))

for ; ackIterator.Valid(); ackIterator.Next() {
    store.Delete(ackIterator.key)
}


receiptIterator := store.ReverseIterator(store, host.PacketReceiptKey(sourceChain, destChain, 0), host.PacketReceiptKey(sourceChain, destChain, sequence+1))

for ; receiptIterator.Valid(); receiptIterator.Next() {
    store.Delete(receiptIterator.key)
}


```

File: modules/tibc/core/04-packet/keeper/keeper.go [hunk 7]
```diff
@@ -305,3 +326,37 @@ func (k Keeper) iterateHashes(_ sdk.Context, iterator db.Iterator, cb func(sourc
...
 }
+
+func (k Keeper) ValidateCleanPacket(ctx sdk.Context, sourceChain, destChain string, sequence uint64) error {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte(host.PacketCommitmentPrefixPath(sourceChain, destChain)))
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		seq := binary.BigEndian.Uint64(iterator.Key())
+		if sequence < seq {
+			break
```
Comment (chengwenxi) [inline modules/tibc/core/04-packet/keeper/keeper.go:337]: return error

**Review by chengwenxi — APPROVED** [review]

File: modules/tibc/core/04-packet/keeper/keeper.go [hunk 8]
```diff
@@ -1,6 +1,8 @@
 package keeper
 
 import (
+	"encoding/binary"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
```
Comment (dreamer-zq) [inline modules/tibc/core/04-packet/keeper/keeper.go]: code  format

File: modules/tibc/core/04-packet/keeper/keeper.go [hunk 9]
```diff
@@ -126,6 +134,19 @@ func (k Keeper) deletePacketCommitment(ctx sdk.Context, sourceChain, destChain s
 	store.Delete(host.PacketCommitmentKey(sourceChain, destChain, sequence))
 }
 
+// GetPacketCommitment gets the packet commitment hash from the store
```
Comment (dreamer-zq) [inline modules/tibc/core/04-packet/keeper/keeper.go:138]:
```suggestion
// GetCleanPacketCommitment gets the packet commitment hash from the store
```

File: modules/tibc/core/04-packet/keeper/keeper.go [hunk 10]
```diff
@@ -126,6 +134,19 @@ func (k Keeper) deletePacketCommitment(ctx sdk.Context, sourceChain, destChain s
...
+func (k Keeper) GetCleanPacketCommitment(ctx sdk.Context, sourceChain, destChain string) []byte {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(host.CleanPacketCommitmentKey(sourceChain, destChain))
+	return bz
+}
+
+// SetPacketCommitment sets the packet commitment hash to the store
```
Comment (dreamer-zq) [inline modules/tibc/core/04-packet/keeper/keeper.go:145]:
```suggestion
// SetCleanPacketCommitment sets the packet commitment hash to the store
```

File: modules/tibc/core/04-packet/keeper/packet.go [hunk 11]
```diff
@@ -255,7 +264,18 @@ func (k Keeper) AcknowledgePacket(
...
+	var targetClientID string
+	if packet.GetSourceChain() == k.clientKeeper.GetChainName(ctx) {
+		if len(packet.GetRelayChain()) > 0 {
+			targetClientID = packet.GetRelayChain()
+		} else {
+			targetClientID = packet.GetDestChain()
+		}
+	} else {
+		isRelay = true
+		targetClientID = packet.GetDestChain()
```
Comment (dreamer-zq) [inline modules/tibc/core/04-packet/keeper/packet.go]:
```suggestion
		targetClientName = packet.GetDestChain()
```

**Review by dreamer-zq — APPROVED** [review]

**Pull request merged**